src/dmd/dtoh.d

```diff
@@ -387,6 +389,106 @@ public:
+    }
+
+    private void writeIdentifier(const Identifier ident, const Loc loc, const char* kind)
+    {
+        void warnCxxCompat(const(char)* reason)
+        {
+            static bool warned = false;
+            warning(loc, "%s `%s` is a %s", kind, ident.toChars(), reason);
+
+            if (!warned)
+            {
+                warningSupplemental(loc, "The generated C++ header will contain " ~
+                                    "identifiers that are keywords in C++");
+                warned = true;
+            }
+        }
+
+        if (global.params.warnings != DiagnosticReporting.off)
+        {
+            // Warn about identifiers that are keywords in C++.
+            switch (ident.toString())
+            {
+                // C++ operators
+                case "and":
+                case "and_eq":
+                case "bitand":
+                case "bitor":
+                case "compl":
+                case "not":
+                case "not_eq":
+                case "or":
+                case "or_eq":
+                case "xor":
+                case "xor_eq":
+                    warnCxxCompat("special operator in C++");
+                    break;
+
+                // C++ keywords
+                case "const_cast":
+                case "delete":
+                case "dynamic_cast":
+                case "explicit":
+                case "friend":
+                case "inline":
+                case "mutable":
+                case "namespace":
+                case "operator":
+                case "register":
+                case "reinterpret_cast":
+                case "signed":
+                case "static_cast":
+                case "typedef":
+                case "typename":
+                case "unsigned":
+                case "using":
+                case "virtual":
+                case "volatile":
+                    warnCxxCompat("keyword in C++");
+                    break;
+
+                // C++11 keywords
+                case "alignas":
+                case "alignof":
+                case "char16_t":
+                case "char32_t":
+                case "constexpr":
+                case "decltype":
+                case "noexcept":
+                case "nullptr":
+                case "static_assert":
+                case "thread_local":
+                    if (global.params.cplusplus >= CppStdRevision.cpp11)
+                        warnCxxCompat("keyword in C++11");
+                    break;
+
+                // C++20 keywords
+                case "char8_t":
+                case "consteval":
+                case "constinit":
+                // Concepts-related keywords
+                case "concept":
+                case "requires":
+                // Coroutines-related keywords
+                case "co_await":
+                case "co_yield":
+                case "co_return":
+                    version (none)
```

Comment:
Since I raised this, we now have C++20 std in the enum.

Reply:
See #12118